pkg/assembler/backends/inmem/artifact.go

```diff
@@ -81,11 +81,24 @@ func (n *artStruct) setCertifyBadLinks(id uint32) { n.badLinks = append(n.badLin
 // Ingest Artifacts
 
 func (c *demoClient) IngestArtifact(ctx context.Context, artifact *model.ArtifactInputSpec) (*model.Artifact, error) {
+	return c.ingestArtifact(ctx, artifact, true)
+}
+
+func (c *demoClient) ingestArtifact(ctx context.Context, artifact *model.ArtifactInputSpec, readOnly bool) (*model.Artifact, error) {
 	algorithm := strings.ToLower(artifact.Algorithm)
 	digest := strings.ToLower(artifact.Digest)
-	a, err := c.artifactByKey(algorithm, digest)
 
+	c.lock(readOnly)
+	defer c.unlock(readOnly)
+
+	a, err := c.artifactByKey(algorithm, digest)
 	if err != nil {
+		if readOnly {
+			c.m.RUnlock()
+			a, err := c.ingestArtifact(ctx, artifact, false)
+			c.m.RLock() // relock so that defer unlock does not panic
```

Comment:
hmm i think we can do some other type of manipulation that is less costly here than acquiring the lock again, maybe something along the lines of https://stackoverflow.com/questions/50318473/is-there-a-way-to-manipulate-the-defer-stack

Reply:
this sounds fine for now, but may need to revisit if we run into concurrency perf issues